source/common/stats/stat_merger.cc

```diff
@@ -0,0 +1,93 @@
+#include "common/stats/stat_merger.h"
+
+#include <regex>
+
+namespace Envoy {
+namespace Stats {
+
+StatMerger::StatMerger(Stats::Store& target_store) : target_store_(target_store) {}
+
+void StatMerger::mergeStats(const Protobuf::Map<std::string, uint64_t>& counters,
+                            const Protobuf::Map<std::string, uint64_t>& gauges) {
+  for (const auto& counter : counters) {
+    uint64_t new_parent_value = counter.second;
+    auto found_value = parent_counter_values_.find(counter.first);
+    uint64_t old_parent_value =
+        found_value == parent_counter_values_.end() ? 0 : found_value->second;
+    parent_counter_values_[counter.first] = new_parent_value;
+    target_store_.counter(counter.first).add(new_parent_value - old_parent_value);
+  }
+
+  // Gauge name *substrings*, and special logic to use for combining those gauges' values.
```

Comment:
I'm sorry that I'm just asking about this now, but the more that I look at this the more concerned I am about this being fragile as we add new gauges. We can perhaps leave this to a followup/TODO, but what your thoughts on somehow making the combination logic be required when a gauge is defined? Do you think that would improve things and have the logic be closer to the code? That seems better to me. WDYT?

Reply:
No, I totally know what you mean; it's been nagging at me the whole time. It needs to be impossible for someone used to shared memory stats and unaware of these changes to add a gauge that might not be simply accumulatable.

I haven't added anything like that because it had seemed like the only real way to mechanically require specification of the logic would be to mess with the stat declaration macros, which might be messy (both for work in this PR and the macros' cleanliness going forward). But, now that I'm thinking about it again, here's a new idea: what if we have some gauge registry where you're required to specify your gauge's combination logic, even if it's just accumulation, with a test mechanism that would run in the integration tests (which presumably cause all stats to be instantiated at some point).

(I'm suggesting this as a new separate thing because just using the existing list of exceptions would make the periodic combination be O(n^2) in number of gauges).

Reply:
Oh actually, here's a slightly cleaner approach: do have this check be in with the combination stuff, but only do the check after first checking the actual exceptions, *and* cache the results. Or, maybe better yet, do the check just when a stat is first allocated. Either way, you don't need a separate test mechanism.

Reply:
I like your idea of having some type of registration mechanism that we then test and it fails if a gauge has not been added (or something like that). I think my feeling is we should ship this since it's a huge improvement, and add a TODO and file an issue to come back to the registration stuff? That would allow us to tackle that in a smaller change and @jmarantz might also have some thoughts. WDYT?

Reply:
I agree with leaving a TODO. There should be currently no stats missed. If this is merged without that mechanism, then in the long run, a messed-up-by-hot-restart stat or two might be added. However, once the mechanism is added, we can just run the "which gauges use set()" audit from #5790 again, and inspect any new ones. After that, the problem would be fully solved.